### PR TITLE
Wrap MastForest in Program and Library in Arc

### LINKED ISSUE #1430: Minor refactoring of the `Library` struct
Current `Library` struct looks like so:
```rust
pub struct Library {
    digest: RpoDigest,
    exports: BTreeMap<QualifiedProcedureName, Export>,
    mast_forest: MastForest,
}

enum Export {
    Local(MastNodeId),
    External(RpoDigest),
}
```

There are a couple of things we may want to change here:

1. We are missing some metadata from library definition. One example of this is `version` but there are also other potential things we may want to track about procedures and the library itself. Regarding `version`, we should decide if we actually need it, and if not, remove the `Version` struct from the codebase.
2. We can also consider removing `Export` enum because information about whether a given procedure is local or re-exported can be inferred from the underlying `mast_forest`.

### PR DESCRIPTION
**Pull request by bobbinth**
This PR wraps `MastForest` fields in `Program` and `Library` structs in `Arc`. This way, cloning these objects becomes a relatively cheap operation.

Additionally, this PR contains the following changes/refactorings:

1. A check was added to make sure that an entrypoint of a `Program` is registered as a root in the `MastForest`. This was already happening - but now we have an explicit check for this.
2. For every re-exported procedure, we now add an `External` node to the `MastForest`. This makes `Library` and `Program` structs a bit more consistent and also removes a somewhat un-intuitive situations where if a library had only re-exported procedures, its `MastForest` would be completely empty. Since now there is an easy way to track re-exported procedures in the `MastForest` itself, the distinction between local and re-exported procedures was removed from the `Library` struct (but I added a method that allows checking if a given export is a re-export).
3. In the `MastForestBuilder`, we now insert `Procedure` structs for re-exported procedures too. This allowed removing `procedure_hashes` map and unify the way we treat local and re-exported procedures.

There is one thing we can improve in the future: if we re-export a procedure from the library and a procedure with the same digest is locally defined, we may end up with some unreachable nodes in the MAST forest. This is technically not a bug - but it does create bloat in the MAST forest. In general, I think if a library has a locally-defined procedure and also re-exports a procedure with the same root, the re-exported procedure should take precedence.

This PR closes #1430. As suggested by @bitwalker, I did not to add `Version` to the `Library`. We should probably either remove `Version` entirely or move it to `miden-core`. But I left this for future PRs.

### CODE REVIEW

File: assembly/src/assembler/mod.rs [hunk 1]
```diff
@@ -493,15 +496,22 @@ impl Assembler {
...
                         &proc_alias.target().into(),
                         &pctx,
                         mast_forest_builder,
                     )?;
+
+                    // insert external node into the MAST forest for this procedure; if a procedure
+                    // with the same MAST rood had been previously added to the builder, this will
+                    // have no effect
+                    let proc_node_id = mast_forest_builder.ensure_external(proc_mast_root)?;
+                    let procedure = pctx.into_procedure(proc_mast_root, proc_node_id);
```
Comment (plafer) [inline assembly/src/assembler/mod.rs:510]: `into_procedure()`'s docs specify:

> The passed-in `mast_root` defines the MAST root of the procedure's body while `mast_node_id` specifies the ID of the procedure's body node in the MAST forest in which the procedure is defined.

But here, `mast_node_id` refers to an external node. I think we should update the docs there to clarify that the "body" can also be an `ExternalNode`, which is what re-exported procedures get compiled down to.

**Review by plafer — COMMENTED** [review]
LGTM! Left a few nits

**Review by plafer — APPROVED** [review]

File: assembly/src/library/mod.rs [hunk 2]
```diff
@@ -110,9 +105,30 @@ impl Library {
...
+        self.exports
+            .get(proc_name)
+            .map(|&node_id| self.mast_forest[node_id].is_external())
+            .unwrap_or(false)
+    }
+
     /// Returns the inner [`MastForest`].
-    pub fn mast_forest(&self) -> &MastForest {
-        &self.mast_forest
+    pub fn mast_forest(&self) -> Arc<MastForest> {
```
[the remark this answers is not in the document]
Comment (bobbinth) [reply assembly/src/library/mod.rs]: Good idea! Changed!

File: stdlib/src/lib.rs [hunk 3]
```diff
@@ -22,22 +24,20 @@ impl From<StdLibrary> for Library {
...
-    }
-}
-
 impl StdLibrary {
+    /// Serialized representation of the Miden standard library.
     pub const SERIALIZED: &'static [u8] =
         include_bytes!(concat!(env!("OUT_DIR"), "/assets/std.masl"));
+
+    /// Returns a reference to the [MastForest] underlying the Miden standard library.
+    pub fn mast_forest(&self) -> Arc<MastForest> {
```
[the remark this answers is not in the document]
Comment (bobbinth) [reply stdlib/src/lib.rs]: Changed here as well.

File: assembly/src/assembler/mod.rs (hunk 1 above)
Comment (bobbinth) [reply assembly/src/assembler/mod.rs:510]: Agreed! Added more info to the description.

**Pull request merged**